### PR TITLE
Give IC2 cells container items as GT cells do

### PR DESCRIPTION
**Pull request by ghostflyby**
Give it a GregTech like behavior.

### CODE REVIEW

**Review by eigenraven — CHANGES_REQUESTED** [review]
The solution is to fix EnderIO's fluid handling code to use the item fluid registry correctly, not implement the other fluid interface on the IC2 cells.

**Comment by ghostflyby** [comment]
> The solution is to fix EnderIO's fluid handling code to use the item fluid registry correctly, not implement the other fluid interface on the IC2 cells.

You are right about the interface, it is completely unnecessary.

However, there is nothing wrong in EnderIO, as the consumption of cells is (at least “was”) intended in the design of IC2 and all other containers do the same on cells.

New solution has been pushed, adding a container item should be enough.

**Comment by eigenraven** [comment]
See <https://github.com/GTNewHorizons/Railcraft/pull/28/files#diff-ccf3f1b777563f1ce8505dea8b0f34817b6079a50e12298aca1ab97efbd77f24> for a fix I did for railcraft to not empty IC2 cells. ContainerItem is used for crafting recipes, it's an unrelated api to fluids - `FluidContainerRegistry.drainFluidContainer` returns the empty ic2 cell.

**Comment by ghostflyby** [comment]
I’m reopening this for several reasons.

1. Simplify the fix. `getContainerItem` has been used in most mods for tanks to determine the container, otherwise changes will be needed everywhere (even more mixins in closed-source mods).
2. Sync with GT behavior. GT cells always gives back empty cells without any fixes, the only difference concerned is that gt cells have a ContainerItem of empty cell.
3. least impact on other items. Items are there like Forestry capsules, which will be affected by migrating to `FluidContainerRegistry.drainFluidContainer`, causing potential game balance change.

**Comment by eigenraven** [comment]
This is still a breaking change to all IC2 crafting recipes using filled cells, most of them are intended to consume the cell (e.g. the vanilla IC2 obsidian recipe) as a balancing factor. Imo, we shouldn't work around broken fluid code by modifying crafting recipes, we should be fixing the broken fluid code. The fluid container registry supports items that get destroyed on draining too (like forestry capsules).

**Comment by miozune** [comment]
I think this issue has been addressed by https://github.com/GTNewHorizons/EnderCore/pull/17

**Comment by ghostflyby** [comment]
> I think this issue has been addressed by [GTNewHorizons/EnderCore#17](https://github.com/GTNewHorizons/EnderCore/pull/17)

Yes, but I leave this open only because GT cells have a containerItem and IC2 cells don’t.

**Comment by miozune** [comment]
GT having container has nothing to do with IC2. As raven said, it's design choice.

**Comment by ghostflyby** [comment]
EnderIO Tanks’ eating IC2 cells was a design choice of both mods either.

**Comment by miozune** [comment]
- EnderIO / EnderCore: It was using unrelated API to drain fluid, not only for EIO tank but also for generic fluid manipulation. It's clearly a bug.
- I loaded up IC2 1.12.2 (industrialcraft-2-2.8.222-ex112.jar), which was released on Jun 19, 2022, and cells are still consumed with obsidian crafting recipe, despite this behavior has been widely known for years. So we can guess it's intended behavior.

**Comment by ghostflyby** [comment]
> The fluid container registry supports items that get destroyed on draining too (like forestry capsules).
forestry capsules are not being destroyed in your railcraft fix (and my EnderCore fix), so is this a expected behavior or a unexpected balance issue?

**Comment by ghostflyby** [comment]
> * EnderIO / EnderCore: It was using unrelated API to drain fluid, not only for EIO tank but also for generic fluid manipulation. It's clearly a bug.
> * I loaded up IC2 1.12.2 (industrialcraft-2-2.8.222-ex112.jar), which was released on Jun 19, 2022, and cells are still consumed with obsidian crafting recipe, despite this behavior has been widely known for years. So we can guess it's intended behavior.

The 1.12.2 fluid API is far more complete. In 1.7.10, all mods tanks expecting to identify between one-time and reusable containers use `getContainerItem`. I have spotted such behavior in extra utilities, railcraft, IC2, EnderIO and Forestry when handling fluids, this is clearly a convention.

File: src/main/java/com/mitchej123/hodgepodge/mixins/late/ic2/MixinIC2ItemCell.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+@Mixin(value = ItemCell.class, remap = false)
+abstract public class MixinIC2ItemCell extends ItemIC2 {
+
+    public MixinIC2ItemCell(InternalName internalName) {
+        super(internalName);
+    }
+
+    @Inject(method = "<init>", at = @At("RETURN"))
+    public void hodgepodge$init(CallbackInfo ci) {
+        setContainerItem(Ic2Items.cell.getItem());
```
Comment (miozune) [inline src/main/java/com/mitchej123/hodgepodge/mixins/late/ic2/MixinIC2ItemCell.java:22]: This should not run on empty cell creation

**Review by miozune — COMMENTED** [review]
On second thought, now I think this change would be reasonable. Item container is used not only for crafting recipes but also others like TC infusion. Having consistent behavior as GT cells or vanilla buckets makes sense.
However, we should not consider this as correct fix for fluid container issue. Other mods might also have items with similar behavior, and if we find a mod that misuse item container, it should still be corrected to use fluid container registry.

File: src/main/java/com/mitchej123/hodgepodge/mixins/late/ic2/MixinIC2ItemCell.java (hunk 1 above)
Comment (ghostflyby) [reply src/main/java/com/mitchej123/hodgepodge/mixins/late/ic2/MixinIC2ItemCell.java]: All cells are sharing the same Item instance, so I override the methods instead.

**Review by miozune — APPROVED** [review]

**Comment by eigenraven** [comment]
Hodgepodge is also used by other modpacks, so maybe this tweak should be off by default except in GTNH?

**Review by eigenraven — APPROVED** [review]

**Comment by Alexdoru** [comment]
> Hodgepodge is also used by other modpacks, so maybe this tweak should be off by default except in GTNH?

@ghostflyby  is someone making a PR to add a default config to true on the modpack repo ?

**Comment by Dream-Master** [comment]
@Alexdoru done by https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/pull/14704

**Pull request merged**